### ISSUE TITLE
Additional Conformance class with requirements about number of 'reduced-resolution-subfiles'?

### ISSUE THREAD

**Issue by cholmes** [opengeospatial/CloudOptimizedGeoTIFF#24]
The COG spec is looking good, and the '[reduced resolution subfile](https://portal.ogc.org/files/102116#_requirement_reduced_resolution_subfiles)' is a welcome addition, to help clients determine if there are overviews/subfiles. 

But I don't see anything that actually gives guidance on how many subfiles a given file should have. Right now you can have a COG with no overviews, and many clients trying to render on the web will choke or be too slow. It seems like even with the reduced resolution subfile conformance class someone could have that conformance class on a huge file, with only one subfile, and it'd meet the requirement but still not be that useful for web rendering.

I think it'd be a big win for COG to have an additional conformance class where any visualization software (javascript front-end reading directly, tile server in between, Google Earth Engine, etc) can know that files that meet the new conformance requirement will give a good experience to users. See https://twitter.com/tmcw/status/1461068163529981973 for a representative discussion on this issue (it comes up fairly often). Or see [this example](https://www.cogeo.org/map/#/url/https%3A%2F%2Fstorage.googleapis.com%2Fopen-cogs%2Fplanet-stac%2Ffusion%2FUTM-24000%2F14N%2F29E-188N%2FPF-SR%2F2021-01-13.tif/center/-96.5326,40.8407/zoom/12) the file that slowly roads is a COG, but has zero overviews (and just one single overview would help some but it'd still not be as snappy as well-overviewed COG's)

I don't see this going quite as far as the [web-optimized COG](https://cogeotiff.github.io/rio-cogeo/Advanced/#web-optimized-cog) (which could be an additional conformance class to add). I see it as an 'in-between' that many would be happy to implement - keep their projection (ie don't require web-mercator), perhaps tiles smaller than 1024 or 512, and power of 2 overviews until you get to the tile size. 

Basically write the requirement to be what gdal cog driver and `rio cogeo` create by default (or a bit looser than them), where they create lots of nice overviews and tile things up for you.

cc @ogcscotts & @tschaub

**Comment by ogcscotts** [comment]
@cholmes it seems that this conformance class could be written in a way not to break backward compatibility with existing COG data.

**Comment by cholmes** [comment]
I believe so. A good number of existing COG's should be able to meet the conformance class (like anything made by the defaults of rasterio or gdal). But it'd be very helpful to be able to 'validate' any COG, and say if it meets this conformance class. Then clients could say 'I only accept COG's that meet it'.

**Comment by tschaub** [comment]
Ideally there would be a conformance class that

 * imposed a max `TileWidth` and `TileLength` of 512 (or maybe 1024)
 * required a reduced resolution subfile for each resolution in the sequence $r_n = 2^n r_0$ until the lowest resolution subfile had 1 `TilesAcross` and 1 `TilesDown` (where $r_0$ represents the number of map units per pixel of the base image) 

On top of that, it would increase interoperability to require that `TileWidth` equals `TileLength` and that the two were the same for all reduced resolution subfiles (and the base image).

I'm not sure there is much utility in a conformance class that just requires one or more reduced resolution subfile (and nothing else).  The tiling requirement is what makes COG reliable for (predictably sized) reads at the base image resolution.  Rendering an overview with COG is a gamble unless there is some guarantee about the structure of the reduced resolution subfiles (this only becomes reliable if they are distributed at a range of resolutions and if the lowest resolution one is reasonably sized (e.g. one tile)).

**Comment by joanma747** [comment]
The new requirements class "Requirement Class Efficient GeoTIFF Overviews" addresses this issue. I'm trying to capture the 2 requirements proposed by @tschaub avoiding imposing any particular tile size or "powers of 2 resolutions". I hope that is still acceptable.

**Comment by rouault** [comment]
> avoiding imposing any particular tile size or "powers of 2 resolutions". I hope that is still acceptable.

I believe the interest of the extra (optional) conformance class is to define a simple and predicatable sub-profile of COGs, so probably that requiring power of 2 is a reasonable expectation.

There's however a difficulty related to resolution in COG, if the size in pixels in a dimension is not a multiple of 2. For example let's say the full resolution is 2049x1025 pixels, and the first overview is 1024x512 pixels. Technically the resolution difference is 2.0009765625 x 2.001953125.
So probably this profile should also mandate that the ImageWidth and ImageLength of the full resolution IFD are a multiple of 2^n.

**Comment by cholmes** [comment]
Yeah, I agree that we need to impose some guidelines to make it 'work'. I could see the 'powers of 2 resolution' being a _minimum_ of some sort - like it doesn't need to be exactly that, but should at least. Maybe that also helps with the issue even raises?

**Comment by rouault** [comment]
There's also probably a presentation / formal issue with how it is presented. Currently it is a paragraph "7.3. Requirement Class Efficient GeoTIFF Overviews", inserted between "7.2. Requirement Class GeoTIFF Overviews" and "7.4. Requirement Class GeoTIFF Keys". It is not obvious at all that 7.3 is *optional*, not something we require for a valid COG. This should probably be in a top level paragraph and with some title clearly showing this is an optional profile (especially if we require ImageWidth/ImageLength to be multiple of 2^n - to have overview levels strictly varying by a resolution factor of 2 - which is not something that  the GDAL COG driver will do by default), to be applied in addition of the base requirements of paragraph 7.

**Comment by geospatial-jeff** [comment]
The choice of tile size for each subfile, number of subfiles, and relationship between resolution across subfiles (ex. `2^n`) is so use-case driven that I don't think it makes sense to add a conformance class restricting what COGs should and shouldn't do.  I think the important things are mostly already covered by the spec but maybe it could be more explicit with how they apply to subfiles:
- Subfiles are optional but SHOULD exist if the COG is intended to be visualized over the web.
- Subfiles should have square tiles, and preferably multiples of 16.
- IFDs for each subfile should be ordered properly.

It makes sense to add recommended best practices for specific COG profiles (ex. web-optimized) but this should not be a conformance class as it is unrealistic that one conformance class can cover all use cases.

**Comment by tschaub** [comment]
> It makes sense to add recommended best practices for specific COG profiles (ex. web-optimized) but this should not be a conformance class

Is there a plan to have "profiles" in addition to "conformance classes?"  My understanding is that the conformance classes are essentially different profiles.

The way I see it, the purpose of a "web optimized" conformance class (by this or any other name) would be to signal to clients that a COG is suitable for predictable and reasonable sized reads across all resolutions (this occurs with reasonably sized tiles aligned to a regular tile grid starting with one at the top and continuing to the "full" resolution).

It doesn't put any additional burden on data providers to have this conformance class in the spec.  Data providers that do not wish to conform to the additional requirements don't need to (they only need to conform with one or more of the other conformance classes).

I think the "web optimized" (or Efficient Overview) conformance class is most useful as a signal to clients.  In addition, I think having many conformance classes ends up reducing the potential value of an optimized/efficient conformance class.  If there were fewer conformance classes, I think the result would be fewer flavors of COG in the wild (in the future) - and ideally a larger portion of them would be optimized/efficient for use on the web.

**Comment by geospatial-jeff** [comment]
> Is there a plan to have "profiles" in addition to "conformance classes?" My understanding is that the conformance classes are essentially different profiles.

I'm not so sure as there are some conformance classes (like `Requirement Class GeoTIFF Tiles`) that should apply to all COGs regardless of whether or not the image is web-optimized (or follows any other "profile").  I would consider this a "conformance class" because it is a requirement that all COGs should satisfy while including web-friendly overviews is a "flavor/profile" of COG that is useful for some use cases (web visualization) but not a requirement of others (ex. machine learning workflows).


> The way I see it, the purpose of a "web optimized" conformance class (by this or any other name) would be to signal to clients that a COG is suitable for predictable and reasonable sized reads across all resolutions (this occurs with reasonably sized tiles aligned to a regular tile grid starting with one at the top and continuing to the "full" resolution).

Maybe the term "web optimized" is overloaded?  In `rio-cogeo` it refers to a COG where each internal tile in the image aligns with a tiling scheme (typically XYZ/web mercator) - 256x256 w/ matching bounds and resolution of the appropriate XYZ tile.  Such that a client or tile server reading the COG only has to send one HTTP request to fetch a single XYZ tile.  I wouldn't call any COG with overviews and `2^n` decimation "web optimized".

I agree with this purpose and, as I said in my previous comment, I think these requirements are for the most part already included in the spec although it could be more explicit how they apply to overviews/subfiles.

> I think having many conformance classes ends up reducing the potential value of an optimized/efficient conformance class. If there were fewer conformance classes, I think the result would be fewer flavors of COG in the wild (in the future) - and ideally a larger portion of them would be optimized/efficient for use on the web.

I agree with this statement in theory; but in practice there is not really a simple definition of a COG that is optimized/efficient for all uses on the web because it is so use-case driven.  The variety of COGs that we see today in the wild is evidence of this.  If there was a clear-cut winner for how to optimize the internal structure of a COG for use on the web it would have been widely adopted already through various implementations but that is far from the case.

TIFF (for better or for worse) is a very very flexible file format.  COG adds some constraints on top of the TIFF spec to optimize access patterns for the cloud, but is still very flexibile in its own right.  My opinion is the spec needs to walk a very narrow line that describes these constraints without over-prescribing solutions.

**Comment by cholmes** [comment]
So I'm the one who originally proposed this as a 'conformance class', and definitely saw it as an 'optional' one, where the others would all be required. I'm not exactly sure on profile vs conformance class, but let me try to state the core goal I've got, that I think everyone is likely aligned to.

1) Core COG Spec should have a number of required conformance classes that correspond to what is in practice today. This makes no requirement about overviews.
2) There should be an optional 'option' (to pick a less overloaded term) that narrows things down a bit - mostly including sufficient overviews that most renderers will have a reasonable chance of rendering it. But not going as far as the rio-cogeo 'web-optimized' profile - no mercator requirements, no 256 tile requirement. This needs a good name, and I agree 'web-optimized' is not right for this.
3) We could could consider another formal 'option' for the rio-cogeo web-optimized one.

For 2 and 3 my goal is that providers can say 'I implement this option', and that it's pretty easy for clients to check. And any rendering (pure javascript client side, server processing engine (GEE, sentinel hub) or tile server) that can understand projections can try to use it 2), and 3) anything that can do web mercator is fine.

I do feel like the web-optimized COG's have been less used, because they really optimize for one purpose, so I'm less attached to them. The big 'opportunity' to me feels like 2), which aligns with your point @geospatial-jeff that the overviews SHOULD be a requirement. We don't try to make it a new requirement, but we make it something that is easy to check.

I think we maybe need some advice on the best OGC way to do things. My understanding had been the conformance classes are like the core building blocks, and some are required, some are optional. I forget when you're supposed to use 'profiles', etc. But I think at the very least the core spec should make clear 'the core COG is X required classes), and the 'efficient overviews' is an option on top of it.

**Comment by cholmes** [comment]
And one more point:
>  If there was a clear-cut winner for how to optimize the internal structure of a COG for use on the web it would have been widely adopted already through various implementations but that is far from the case.

**Comment by cholmes** [comment]
And one more point: 
>  If there was a clear-cut winner for how to optimize the internal structure of a COG for use on the web it would have been widely adopted already through various implementations but that is far from the case.

I do think there is some significant adoption of 'including sufficient overviews', of powers of 2, as it's the default of GDAL and rasterio cogeo. So I think it's important that we have a way to say 'you have a valid 'efficient overview' COG' or you don't. Saying that it doesn't implement the optional conformance class is the core for that.

But I do think we need a solid 'name' for COG's with overviews.

**Comment by rouault** [comment]
> But I do think we need a solid 'name' for COG's with overviews.

"pyramidal COG" ? (https://openlayers.org/en/latest/examples/cog-pyramid.html)

**Comment by tschaub** [comment]
> Is there a plan to have "profiles" in addition to "conformance classes?"

This was meant to be more of a rhetorical question.  I sincerely hope that there will *not* be "profiles" (or anything else representing a distinct COG flavor) in addition to "conformance classes."  Please do not add to the proliferation of options for optimizing GeoTIFFs for the cloud.  Formalizing and tightening up the specification is good.  Adding flavors/options/profiles/conformance classes etc. hampers interoperability by adding to the variety that clients are expected to deal with.

> I think we maybe need some advice on the best OGC way to do things. My understanding had been the conformance classes are like the core building blocks, and some are required, some are optional. I forget when you're supposed to use 'profiles', etc.

So we're all on the same page, the current draft spec says an implementation only needs to conform with one or more of the conformance classes.  See: https://github.com/opengeospatial/CloudOptimizedGeoTIFF/blob/79589f39d0d9df68d6cd8adfb93159ef6cb245cc/standard/clause_2_conformance.adoc#L22-L23

The conformance classes are logical groups of requirements and recommendations.  Importantly, conformance classes can have dependencies on one another.  For example, if you implement the [GeoTIFF Overviews Conformance Class](http://docs.ogc.org/DRAFTS/21-026.html#geotiff-overviews-requirements-class), you must also implement the [GeoTIFF Tiles Conformance Class](http://docs.ogc.org/DRAFTS/21-026.html#geotiff-format-requirements-class) (maybe should be called "Format" or "Core" instead of "Tiles"?).

The benefit of OGC conformance classes is that they give providers a lot of flexibility.  A provider can choose what level of conformance is appropriate for them, picking and choosing the requirement sets that are appropriate.

This benefit for providers can be a drawback for consumers.  Many conformance classes make it easy for providers with a wide range of use cases to say that they conform with a specification.  This same flexibility makes it hard for a consumer (particularly a generic client library) to say that they conform with a specification, since they must be prepared to handle a wide range of GeoTIFF flavors.

I think the OGC needs to keep this tension in mind when developing specifications.  Lots of conformance classes are nice for providers, but can make life hard for consumers.

Back to the conformance class being discussed here.  I'm going to use the term "Efficient" since "optimized" is already taken and there seems to be baggage associated with "web."  If the OGC COG spec were to have an "Efficient" conformance class, I think the primary benefit would be as a signal to "limited capability" clients¹ that only wanted to work with COGs that provided some sort of reasonability guarantee when reading.  Ideally, providers would be motivated to work with these clients and they would take the additional steps necessary to conform with the requirements.  This doesn't impose any additional burden on other providers that want to use the OGC COG spec but don't care about limited capability clients¹ - they can just implement the requirements of the "Core" conformance class or any other and ignore the "Efficient" requirements.

> "pyramidal COG" ?

Yes, maybe a "Pyramid Conformance Class" instead of "Efficient" would be more descriptive.  But really, not just any pyramid.  A full quad tree please.

¹ I don't actually think talking about "limited capability" clients is a good way to group things.  I think "web" is a better term here really.  For example, a browser and QGIS have very different constraints or capabilities, but may both find it efficient to access a remote GeoTIFF via HTTP using the same patterns (reading reasonably sized tiles and a full range of overviews).  The thing that is common in these cases is probably better described as "web" rather than "limited capability."

**Comment by hobu** [comment]
> But I do think we need a solid 'name' for COG's with overviews.

facetiousness aside, why isn't that name simply 'COG', as in any COG must have overviews as expected to be called a COG?

COG is (or was) an opinionated organization of GeoTIFFs to support specific kinds of tasks. Those opinions were stated by (GDAL) code too, not specification documents. Tim's point about flexibility for producers pressing complexity and performance pain onto data consumers is very important to think about, and if OGC language allows this, that, and the other thing all be *conformant* COGs, it turns it all to mush.

It's the **constraint** of COG that makes it special. Backward compatible *and* streamable over HTTP 206. Both of those things together with some sane behavior choices. Otherwise people should go make a proper streamable raster format without all of the legacy of TIFF and GeoTIFF and bootstrap implementation support for it throughout the industry.

My perspective is this same story exists in point cloud land with [COPC.io](https://copc.io), which deliberately follows the pattern. There's a gazillion ways to organize a point cloud file. COPC has the perspective of **we are doing it this way** and calling it COPC. 

I don't know that I'm adding much to the conversation here other than to confirm my fear that when community efforts like COG start iterating inside of OGC, they get larded up with hopes and dreams instead of being focused on what made them so successful as to bubble out in the first place.

**Comment by rouault** [comment]
> as in any COG must have overviews as expected to be called a COG?

I believe the rationale for not making overviews a requirement is that you don't need overviews for most analytic use cases. (but you need them for most visualization purposes). The GDAL COG driver has an option to skip overview generation.

**Comment by joanma747** [comment]
Let me clarify the optionality of conformance classes. By definition ALL conformance classes are optional. BUT There is a way to "pack" them: defining dependencies.

This is what we have today: We have 5 conformance classes:
"format" it is the core
"overviews" depends on "format"
"efficient overviews" depends on "overviews that depends on "format"
"geotiff keys" depends on "format"
"httprange" depends on "format"

I'm not completely sure this is the right way of doing it. The original intention was to make "geotiff keys" completely optional but today, after  I believe it makes no sense to do it this way. I propose to change this to:
"format" it is the core
"overviews" depends on "format"
"geotiff keys" should depend on "overviews" that depends on "format"
"efficient overviews" should depend on "geotiff keys" that depends on "overviews" that depends on "format"
"httprange" depends on "format"

So "efficient overviews" will imply "everything" (except "httprange", that is a completely different story)

If I do not hear yelling, all do this change.

"httprange" depends on "format"

**Comment by tschaub** [comment]
@joanma747 - your proposed dependency changes make sense to me.

I think the remaining question is the original reason this issue was opened: how efficient is the Efficient Conformance Class?  If the Efficient requirements allow degenerate cases (a file with a single overview with a resolution factor of 1.00001) to be classified as “efficient” it is a misuse of the term and renders the conformance class useless.

Put another way, what is the downside of having and Efficient Conformance Class that was more opinionated (smallish, uniformly sized tiles, full quad tree)?  As I think we all understand now, having such a conformance class doesn’t impose any burden on providers.  They can choose to satisfy just the core and still comply with the spec.

**Comment by geospatial-jeff** [comment]
> "geotiff keys" should depend on "overviews" that depends on "format"

This doesn't make sense to me.  It implies that an image with geotiff keys must also have overviews but that isn't true.  In fact the presence/absence of overviews and geotiff key aren't really related to each other at all.  It's normal for GeoTIFFs / COGs to only include geotiff keys on the IFD with the full resolution image, they are usually not included on any IFDS for additional overviews / subfiles.  Instead I think the dependency graph should look like:

<img width="473" alt="image" src="https://user-images.githubusercontent.com/40916268/211367708-7c7e7a49-97e5-4455-ba27-b34e19a8cf9a.png">

I'm also not sure if `efficient overviews` should depend on `geotiff keys` for the same reason.  One argument is that geotiff keys enable efficient reading of the underlying data via dynamic tile serves like titiler which rely on these geotiff keys for warping/reprojection etc.  The counter argument is that there are use cases for "efficient overviews" on non-spatial data where geotiff keys are not a requirement.  I've also created COGs with "efficient overviews" but without geokeys which are then served through spatially-enabled standards like XYZ/WMTS but that is a pretty niche use case.

**Comment by cholmes** [comment]
I wonder if we should aim for a zoom call to try to hash through these things? I feel like getting this stuff right is key.

It seems like there's pretty good consensus on making 'efficient overviews' more opinionated.

I would like to talk through what @hobu said above:

>  why isn't that name simply 'COG', as in any COG must have overviews as expected to be called a COG?

I'd be totally for this, but it'd be an increased requirement for 'OGC cog's'. To be honest I _thought_ COG required overviews for its first couple years of existence, until people started making non-overview ones. I do think if we go this route then we need to have a name for COG's without overviews - the ones that were totally valid before this becomes an OGC spec, that validators today will say 'this is a cog' (though they do give a warning. I think we could call them like 'analytic COG's' or something.

**Comment by joanma747** [comment]
I believe @geospatial-jeff proposal is sensible and I'll implement it.

The proposal from @hobu make me think it could be convenient to be close to the COG acronym meaning so I propose to chance the name of the requirements class from "efficient overviews" to "optimized geotiff". Optimized is more or less a synonyms of efficient "geotiff" is useful in the name to express the dependency on "geotiff.

About the proposal to have a zoom meeting, please consider attending the Monday, February 20th session at 4:10pm - 5:00pm to discuss this and other final aspects of the COG before request the TC for the final voting for an OGC standard

**Comment by joanma747** [comment]
After proposing a possible solution in this, and waited for a while to see if there was no objection and I assumed that the proposed solution was acceptable. I apologize if that was not the case. 

The current status is:
There is a class is called "Optimized GeoTIFF" that requires tiles, overviews and geotiff keys. IMHO it is similar to the "efficient overviews" proposed by @cholmes. In the document overview section, there is a subsection called "Efficient organization of data in a TIFF file" that tries to explain that too.

> For people that can vote, does it make sense to leave any comments on the vote (like "I only approve if we can settle on a useful conformance class for 'efficient' COGs" or something similar)?
This is perfectly possible to vote "yes" with comments. It is not my desire, but it is also possible have "no votes" with comments. 
This comments has to be processed and look for the consensus.

>Can there be changes to the spec after it is approved?
Once it is approved, it is approved as is. We can always release a v1.1 with changes, but it will be a new version.

@ogcscotts will be able to better clarify the process to you.

**Comment by joanma747** [comment]
> Ideally there would be a conformance class that
> 
> * imposed a max `TileWidth` and `TileLength` of 512 (or maybe 1024)
> * required a reduced resolution subfile for each resolution in the sequence rn=2n×r0 until the lowest resolution subfile had 1 `TilesAcross` and 1 `TilesDown` (where r0 represents the number of map units per pixel of the base image)
> 
> On top of that, it would increase interoperability to require that `TileWidth` equals `TileLength` and that the two were the same for all reduced resolution subfiles (and the base image).
> 
> I'm not sure there is much utility in a conformance class that just requires one or more reduced resolution subfile (and nothing else). The tiling requirement is what makes COG reliable for (predictably sized) reads at the base image resolution. Rendering an overview with COG is a gamble unless there is some guarantee about the structure of the reduced resolution subfiles (this only becomes reliable if they are distributed at a range of resolutions and if the lowest resolution one is reasonably sized (e.g. one tile)).

A recommendation has been added in the last version of the document.
